### DISCUSSION TITLE
BSP and JDK 8

### DISCUSSION THREAD

**Discussion by lwronski** [build-server-protocol/build-server-protocol#494] [General]
I would like to open a discussion about dropping support for JDK 8.

After upgrading `jsoniter-scala-core` to `2.13.5` BSP stopped supporting JDK 8 because `jsoniter` from this version doesn't support JDK 8 - [plokhotnyuk/jsoniter-scala#860](https://github.com/plokhotnyuk/jsoniter-scala/pull/860). And this causes the BSP can't be updated in the Bloop - [scalacenter/bloop#1845](https://github.com/scalacenter/bloop/pull/1845).

I'm not sure which solution is better, I can downgrade `jsoniter-scala` in BSP or maybe drop JDK 8 support everywhere, also in bloop.

**Comment by ckipp01** [comment]
Hey @lwronski sorry this has been sitting here for so long without a response. Thanks for bringing this up!

I know on the Metals side we discussed this slightly as well. Currently we still support Java 8 there as you know, as well as in Bloop. So I think that in theory we'd all sort of need to agree and move that forward together. I know @tgodzik had some thoughts there about handling it in Metals similarly to how it's down in scala-cli and just pulling down what we need.

It's unfortunate that `jsoniter` was the deciding factor for this in this repo. Arguably, BSP has already dropped support for JDK 8 due to not currently working. That sort of brings up the following question: Is that what we want?
- If not, we should probably roll `jsoniter` back.
- If not, is `jsoniter` the best choice here?
- If so, how should the rest of the ecosystem respond and how can we align there

I think this is something we'll need some wider input on from other users of BSP.

**Comment by plokhotnyuk** [comment]
I think that rolling back to latest 2.13.5.x would be the simplest option. In case of any critical error I'm able to fix and cut next one version for this branch.

If support of JDK 8 worth it then I can try to build a multi-release JAR for `jsoniter-scala-coreJVM` module. But currently I don't know how to cook it with SBT. I see that [`sbt-multi-release-jar`](https://github.com/sbt/sbt-multi-release-jar) was able to do it for Java 8/9. Would anybody help me adopt it for Java 8/11?

**Comment by lefou** [comment]
In Mill we use BSP to provide an interface for e.g. IDEs. Implementation wise we use bsp4j, which is not using jsoniter, so we are not affected directly, yet. As a build tool, we have to provide support for a rather wide range of environments. So (disregarding my personal thoughts about what the best Java version should be for some random project) I think build tooling still needs to provide support for Java 8. There are still projects requiring that platform, and even when developers tend to use newer platforms, cross compiling for Java 8 with newer JDKs is tricky and risky, so you better run on a Java 8 platform to build for it.
